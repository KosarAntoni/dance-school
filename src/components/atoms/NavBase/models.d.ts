type NavBaseStringNodes = 'url' | 'queryString' | 'className' | 'ariaLabel';

export interface NavBaseProps extends Partial<Record<NavBaseStringNodes, string>> {
  onClick?: () => void;
}
