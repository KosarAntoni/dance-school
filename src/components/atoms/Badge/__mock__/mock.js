export const BadgeDefaultClassesMock = [
  'badge',
  'badge--translucent',
  'badge--orange',
  'test-class',
];

export const BadgeDefaultPropsMock = {
  variant: 'translucent',
  color: 'orange',
  className: 'test-class',
};

export const BadgeChildrenMock = 'test content';
