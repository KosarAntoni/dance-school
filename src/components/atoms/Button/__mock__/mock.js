export const testFunction = jest.fn();

export const buttonDefaultClassesMock = [
  'button',
  'button--dark-green',
  'button--solid',
  'test-class',
];

export const ButtonDefaultPropsMock = {
  onClick: testFunction,
  className: 'test-class',
  variant: 'solid',
  color: 'dark-green',
  size: 20,
};

export const ButtonChildrenMock = 'test content';
