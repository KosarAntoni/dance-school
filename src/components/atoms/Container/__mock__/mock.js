export const ContainerOuterDefaultClassesMock = [
  'container__outer',
  'container--wide',
  'container--gutter',
  'outer-test-class',
];

export const ContainerDefaultClassesMock = ['container', 'test-class'];

export const ContainerDefaultPropsMock = {
  className: 'test-class',
  outerClassName: 'outer-test-class',
  variant: 'wide',
};

export const ContainerChildrenMock = 'test content';
