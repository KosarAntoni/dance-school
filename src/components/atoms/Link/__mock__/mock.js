export const linkDefaultClassesMock = ['link link--solid-blue', 'test-class'];

export const LinkDefaultPropsMock = {
  url: '/about',
  className: 'test-class',
  variant: 'solid',
  color: 'blue',
  size: 20,
  hasArrow: true,
};

export const LinkChildrenMock = 'test content';
