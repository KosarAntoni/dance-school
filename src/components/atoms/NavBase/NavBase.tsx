import React, { FC } from 'react';
import { Link } from 'gatsby';
import { AnchorLink } from 'gatsby-plugin-anchor-links';
import { beginigHashRegexp, beginigSlashRegexp, endingSlashRegexp } from 'shared/regexp';

import { NavBaseProps } from './models.d';

const NavBase: FC<NavBaseProps> = ({
  children = '',
  url,
  queryString,
  ariaLabel,
  onClick,
  className,
  ...rest
}) => {
  const renderDiv = <div className={className}>{children}</div>;

  const renderLink = (): JSX.Element => {
    const isInternal = url && beginigSlashRegexp.test(url);
    const isHashtag = queryString && beginigHashRegexp.test(queryString);

    const linkProps = {
      className,
      'aria-label': ariaLabel,
      onClick,
    };

    if (isInternal && isHashtag) {
      const link = `${url.replace(endingSlashRegexp, '')}${queryString}`;

      return (
        <AnchorLink to={link} {...linkProps} gatsbyLinkProps={rest}>
          {children}
        </AnchorLink>
      );
    }

    if (isInternal) {
      return (
        <Link to={url} {...linkProps} {...rest}>
          {children}
        </Link>
      );
    }

    return (
      <a
        href={url}
        target="_blank"
        rel="noreferrer"
        className={className}
        aria-label={ariaLabel}
        {...rest}
      >
        {children}
      </a>
    );
  };

  const renderButton = (
    <button type="button" className={className} onClick={onClick} aria-label={ariaLabel} {...rest}>
      {children}
    </button>
  );

  const renderComponent = (): JSX.Element => {
    if (url && url !== '#') {
      return renderLink();
    }
    if (onClick) {
      return renderButton;
    }

    return renderDiv;
  };

  return renderComponent();
};

export default NavBase;
