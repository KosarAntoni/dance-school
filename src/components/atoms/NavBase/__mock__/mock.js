export const NavBasePropsMock = {
  testFunction: jest.fn(),
  content: 'test-content',
  internalLink: '/about/',
  ariaLabel: 'test-label',
  queryString: '#test',
  kebabQueryString: '#kebab-test',
  externalLink: 'http://localhost:8000',
  className: 'test-class',
};
