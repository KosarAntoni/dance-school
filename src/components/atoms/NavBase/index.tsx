export * from './models.d';
export { default } from './NavBase';
