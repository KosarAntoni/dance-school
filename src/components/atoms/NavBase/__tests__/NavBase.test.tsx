import React from 'react';
import { Link } from 'gatsby';
import { fireEvent, render, screen } from '@testing-library/react';
import { endingSlashRegexp } from 'shared/regexp';

import { NavBasePropsMock } from '../__mock__/mock';
import NavBase from '../NavBase';

const {
  testFunction,
  content,
  internalLink,
  queryString,
  kebabQueryString,
  ariaLabel,
  externalLink,
  className,
} = NavBasePropsMock;

test('the NavBase component renders correctly with required and optional props', () => {
  // render as internal link
  const { rerender } = render(
    <NavBase url={internalLink} ariaLabel={ariaLabel}>
      {content}
    </NavBase>
  );

  let navBaseNode = screen.getByRole('link');

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(navBaseNode).toMatchSnapshot();

  // the component has href
  expect(navBaseNode).toHaveAttribute('href', internalLink);

  // the component has aria-label
  expect(navBaseNode).toHaveAttribute('aria-label', ariaLabel);

  // the component opens propper page after click
  fireEvent.click(navBaseNode);
  expect(Link).toHaveBeenCalledWith(
    expect.objectContaining({ to: internalLink }),
    expect.anything()
  );

  // rerender as external link
  rerender(<NavBase url={externalLink}>{content}</NavBase>);

  navBaseNode = screen.getByRole('link', { name: content });

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // the component has href
  expect(navBaseNode).toHaveAttribute('href', externalLink);

  // rerender as button
  rerender(<NavBase onClick={testFunction}>{content}</NavBase>);

  navBaseNode = screen.getByRole('button', { name: content });

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // after click right function is called
  fireEvent.click(navBaseNode);
  expect(testFunction).toBeCalled();

  // rerender as blank text
  rerender(<NavBase className={className}>{content}</NavBase>);

  navBaseNode = screen.getByText(content);

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // the component has right class
  expect(navBaseNode).toHaveClass(className);

  // render as internal hash link
  rerender(
    <NavBase url={internalLink} queryString={queryString}>
      {content}
    </NavBase>
  );

  navBaseNode = screen.getByRole('link');

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // the component has href
  const href = `${internalLink.replace(endingSlashRegexp, '')}${queryString}`;
  expect(navBaseNode).toHaveAttribute('href', href);

  // the component opens propper page after click
  fireEvent.click(navBaseNode);
  expect(Link).toHaveBeenCalledWith(expect.objectContaining({ to: href }), expect.anything());

  // render as internal kebab-case hash link
  rerender(
    <NavBase url={internalLink} queryString={kebabQueryString}>
      {content}
    </NavBase>
  );

  navBaseNode = screen.getByRole('link');

  // the component has been rendered
  expect(navBaseNode).toBeInTheDocument();

  // the component has href
  const kebabHref = `${internalLink.replace(endingSlashRegexp, '')}${kebabQueryString}`;
  expect(navBaseNode).toHaveAttribute('href', kebabHref);

  // the component opens propper page after click
  fireEvent.click(navBaseNode);
  expect(Link).toHaveBeenCalledWith(expect.objectContaining({ to: kebabHref }), expect.anything());
});
