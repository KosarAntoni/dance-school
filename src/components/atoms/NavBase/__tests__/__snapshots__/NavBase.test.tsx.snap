// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the NavBase component renders correctly with required and optional props 1`] = `
<a
  aria-label="test-label"
  href="/about/"
>
  test-content
</a>
`;
